### ISSUE TITLE
[Error Report]: System.Private.Xml-related Crash

### ISSUE THREAD

**Issue by TheGamingSiri** [X-Hax/SA-Mod-Manager#92]
Hi, I was just loading up a profile on the new SA2 suite in the mod manager, but I ran into a crash that rendered that section inaccessible. I was able to get back into the SA1 suite by downloading an SA1 mod, but any attempt to enter the SA2 suite again caused the same crash.

I was moving several mods around in the hierarchy, but the game was never loaded and I was saving somewhat frequently as I went. I wasn't able to find the specified XML file in the game's Steam folder, either.

## Exception Report
```
SAMM Information:
	Manager Version: 1.2.0
	Current Game: Sonic Adventure 2
	Game Directory: D:\SteamLibrary\steamapps\common\Sonic Adventure 2
	Manager Location: C:\Program Files (x86)\Steam\steamapps\common\Sonic Adventure DX\
	Running as Admin: False
	Mod Loader Version (Hash): b0dd70904209002b5277ba988a3b456e25b0c17a
```

```
Exception Information:
	Type: InvalidOperationException
	Exception: There is an error in XML document (1, 1).
	Source: System.Private.Xml
	In Function: System.Object Deserialize(System.Xml.XmlReader, System.String, System.Xml.Serialization.XmlDeserializationEvents)
	StackTrace:    at System.Xml.Serialization.XmlSerializer.Deserialize(XmlReader xmlReader, String encodingStyle, XmlDeserializationEvents events)
	   at System.Xml.Serialization.XmlSerializer.Deserialize(Stream stream)
	   at SAModManager.Codes.CodeList.Load(String filename)
	   at SAModManager.MainWindow.LoadModList()
	   at SAModManager.MainWindow.Load(Boolean newSetup)
	   at SAModManager.MainWindow.ComboGameSelection_SelectionChanged(Object sender, SelectionChangedEventArgs e)
	   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state)
	   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
	   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
	   at System.Windows.Threading.DispatcherOperation.InvokeImpl()
	   at MS.Internal.CulturePreservingExecutionContext.CallbackWrapper(Object obj)
	   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
	--- End of stack trace from previous location ---
	   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
	   at MS.Internal.CulturePreservingExecutionContext.Run(CulturePreservingExecutionContext executionContext, ContextCallback callback, Object state)
	   at System.Windows.Threading.DispatcherOperation.Invoke()
	   at System.Windows.Threading.Dispatcher.ProcessQueue()
	   at System.Windows.Threading.Dispatcher.WndProcHook(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
	   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
	   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
	   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
	   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
	   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(DispatcherPriority priority, TimeSpan timeout, Delegate method, Object args, Int32 numArgs)
	   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam)
	   at MS.Win32.UnsafeNativeMethods.DispatchMessage(MSG& msg)
	   at System.Windows.Threading.Dispatcher.PushFrameImpl(DispatcherFrame frame)
	   at System.Windows.Application.RunDispatcher(Object ignore)
	   at System.Windows.Application.RunInternal(Window window)
	   at SAModManager.App.Main()
	InnerException: System.Xml.XmlException: Data at the root level is invalid. Line 1, position 1.
   at System.Xml.XmlTextReaderImpl.Throw(Exception e)
   at System.Xml.XmlTextReaderImpl.Throw(String res, String arg)
   at System.Xml.XmlTextReaderImpl.ParseRootLevelWhitespace()
   at System.Xml.XmlTextReaderImpl.ParseDocumentContent()
   at System.Xml.XmlReader.MoveToContent()
   at Microsoft.Xml.Serialization.GeneratedAssembly.XmlSerializationReaderCodeList.Read9_CodeList()
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

**Comment by Sora-yx** [comment]
The XML file is the Config file that the SA2's official launcher use, you can find it in your **sa2 folder -> Config -> UserConfig.cfg**.
If you can share it here, it would be helpful so we can take a look and maybe see what happened.
Once you shared it here, you should be able to delete it and restart the manager, it will re create it automatically and hopefully fix your issue.

**Comment by TheGamingSiri** [comment]
[UserConfig.zip](https://github.com/X-Hax/SA-Mod-Manager/files/14724366/UserConfig.zip)
Here's the UserConfig.cfg. In the meantime, I reinstalled SA2 to be on the same drive as SA1, but the mod manager broke again shortly thereafter. Enabling all mods and saving seemed to be the trigger.

**Comment by TheGamingSiri** [comment]
UPDATE: I resolved the issue! It turns out 1 of the mods I have is defective. The mod in this case was "Cookie Country Chao Garden." Changing the title or category of this mod causes the entire loader to lock up. How a simple texture swap mod is capable of that, I have no idea, but everything seems to be fine for now. I'll let you know if I find other mods with this problem, but otherwise, this case can be considered closed.

**Comment by TheGamingSiri** [comment]
Last Update: "Improved Waypoints (Chao Garden)" is the only other mod I have discovered that causes the same crash when attempting to add a category to it. With that in mind, I can safely say this can be considered closed, as it doesn't appear to be an issue with the mod loader itself.

**Comment by ItsEasyActually** [comment]
I tried with the first chao mod you mentioned, both a rename and category getting added on the latest release. I was not able to reproduce the behavior of crashing the manager. That behavior is concerning and is something I'd rather confirm is an isolated issue.

Can you detail how you were editing this? Editing mods using the manager's mod editor worked fine and so did manually editing it.

**Comment by TheGamingSiri** [comment]
I originally used the mod manager's editor, which is how I first triggered the crash. The only way to get in at that point was to remove the mod from the mods folder. I then re-downloaded the mod, this time adding a category directly in the mod.ini, but that caused the same crash to the modloader on startup. Confusingly, the crash would keep happening even if I removed the category or undid the name change; the only way to get the mod to work in the loader again was to re-download an entirely new instance of the mod and not tamper with it at all.
